cachix/src/Cachix/Client/Push.hs

```diff
@@ -237,36 +221,87 @@ uploadStorePath cache storePath retrystatus = do
               else Store.validPathInfoDeriver store pathinfo
           deriver <- for deriverPath Store.getStorePathBaseName
           referencesPathSet <- Store.validPathInfoReferences store pathinfo
-          referencesPaths <- sort . fmap toS <$> for referencesPathSet (Store.storePathToPath store)
           references <- sort . fmap toS <$> for referencesPathSet Store.getStorePathBaseName
-          let fp = fingerprint (decodeUtf8With lenientDecode storePathText) narHash narSize referencesPaths
-              sig = case pushParamsSecret cache of
-                PushToken _ -> Nothing
-                PushSigningKey _ signKey -> Just $ toS $ B64.encode $ unSignature $ dsign (signingSecretKey signKey) fp
-              nic =
-                Api.NarInfoCreate
-                  { Api.cStoreHash = storeHash,
-                    Api.cStoreSuffix = storeSuffix,
-                    Api.cNarHash = narHash,
-                    Api.cNarSize = narSize,
-                    Api.cFileSize = fileSize,
-                    Api.cFileHash = toS fileHash,
-                    Api.cReferences = references,
-                    Api.cDeriver = maybe "unknown-deriver" (decodeUtf8With lenientDecode) deriver,
-                    Api.cSig = sig
-                  }
-          escalate $ Api.isNarInfoCreateValid nic
+          return $
+            NarInfo.NarInfo
+              { NarInfo.storePath = toS storePathText,
+                NarInfo.url = "",
+                NarInfo.compression = show $ compressionMethod strategy,
+                NarInfo.fileHash = toS fileHash,
+                NarInfo.fileSize = fileSize,
+                NarInfo.narHash = toS narHash,
+                NarInfo.narSize = narSize,
+                NarInfo.references = Set.fromList references,
+                NarInfo.deriver = decodeUtf8With lenientDecode <$> deriver,
+                NarInfo.system = Nothing,
+                NarInfo.sig = Nothing,
+                NarInfo.ca = Nothing
+              }
+    streamStorePath pushParams narStream makeNarInfo
+
+streamStorePath ::
+  (MonadUnliftIO m) =>
+  PushParams m r ->
+  ConduitT () ByteString (ResourceT m) () ->
+  m NarInfo.SimpleNarInfo ->
+  m r
+streamStorePath pushParams narStream narInfoM = do
+  let cacheName = pushParamsName pushParams
+      store = pushParamsStore pushParams
+      authToken = getCacheAuthToken (pushParamsSecret pushParams)
+      clientEnv = pushParamsClientEnv pushParams
+      cacheClientEnv =
+        clientEnv
+          { baseUrl = (baseUrl clientEnv) {baseUrlHost = toS cacheName <> "." <> baseUrlHost (baseUrl clientEnv)}
+          }
+  storeDir <- Store.storeDir store
+  dummyStorePath <- liftIO $ Store.parseStorePathBaseName "j4fwy5gi1rdlrlbk2c0vnbs7fmlm60a7-coreutils-9.1"
```

Comment:
This brings in old mistakes, we should have storePath passed only for the things that really need it.